src/dd/StateGeneration.cpp

```diff
@@ -62,8 +64,194 @@ vCachedEdge randomNode(Qubit v, vNode* left, vNode* right, Generator& gen,
+        "Insufficient qubit states provided. Requested " + std::to_string(n) +
+        ", but received " + std::to_string(state.size()));
+  }
+
+  vCachedEdge f = vCachedEdge::one();
+  for (std::size_t p = start; p < n + start; ++p) {
+    switch (state[p]) {
+    case BasisStates::zero:
+      f = dd.makeDDNode(static_cast<Qubit>(p),
+                        std::array{f, vCachedEdge::zero()});
+      break;
+    case BasisStates::one:
+      f = dd.makeDDNode(static_cast<Qubit>(p),
+                        std::array{vCachedEdge::zero(), f});
+      break;
+    case BasisStates::plus:
+      f = dd.makeDDNode(static_cast<Qubit>(p),
+                        std::array<vCachedEdge, RADIX>{
+                            {{f.p, dd::SQRT2_2}, {f.p, dd::SQRT2_2}}});
+      break;
+    case BasisStates::minus:
+      f = dd.makeDDNode(static_cast<Qubit>(p),
+                        std::array<vCachedEdge, RADIX>{
+                            {{f.p, dd::SQRT2_2}, {f.p, -dd::SQRT2_2}}});
+      break;
+    case BasisStates::right:
+      f = dd.makeDDNode(static_cast<Qubit>(p),
+                        std::array<vCachedEdge, RADIX>{
+                            {{f.p, dd::SQRT2_2}, {f.p, {0, dd::SQRT2_2}}}});
+      break;
+    case BasisStates::left:
+      f = dd.makeDDNode(static_cast<Qubit>(p),
+                        std::array<vCachedEdge, RADIX>{
+                            {{f.p, dd::SQRT2_2}, {f.p, {0, -dd::SQRT2_2}}}});
+      break;
+    }
+  }
+  const vEdge e{f.p, dd.cn.lookup(f.w)};
+  dd.incRef(e);
+  return e;
+}
+
+VectorDD makeGHZState(const std::size_t n, Package& dd) {
+  suitablePackage(n, dd);
+
+  if (n == 0U) {
+    return vEdge::one();
+  }
+
+  auto leftSubtree = vEdge::one();
+  auto rightSubtree = vEdge::one();
+
+  for (std::size_t p = 0; p < n - 1; ++p) {
+    leftSubtree = dd.makeDDNode(static_cast<Qubit>(p),
+                                std::array{leftSubtree, vEdge::zero()});
+    rightSubtree = dd.makeDDNode(static_cast<Qubit>(p),
+                                 std::array{vEdge::zero(), rightSubtree});
+  }
+
+  const vEdge e = dd.makeDDNode(
+      static_cast<Qubit>(n - 1),
+      std::array<vEdge, RADIX>{
+          {{leftSubtree.p, {&constants::sqrt2over2, &constants::zero}},
+           {rightSubtree.p, {&constants::sqrt2over2, &constants::zero}}}});
+
+  dd.incRef(e);
+
+  return e;
+}
+
+VectorDD makeWState(const std::size_t n, Package& dd) {
+  suitablePackage(n, dd);
+
+  if (n == 0U) {
+    return vEdge::one();
+  }
+
+  auto leftSubtree = vEdge::zero();
+  if ((1. / sqrt(static_cast<double>(n))) < RealNumber::eps) {
+    throw std::runtime_error(
+        "Requested qubit size for generating W-state would lead to an "
+        "underflow due to 1 / sqrt(n) being smaller than the currently set "
+        "tolerance " +
+        std::to_string(RealNumber::eps) +
+        ". If you still wanna run the computation, please lower "
+        "the tolerance accordingly.");
+  }
```

Comment:
@burgholzer How would you force this to throw in a unit test? 🤔 Many thanks!

Reply:
That's not too easy.
Although it should be doable. And I think we actually force that in an existing unit test.
You simply have to raise the default tolerance to a rather high level so that `1/sqrt(n)` becomes smaller than that tolerance.

I'd anyway recommend to simply remove the existing functions to see where the functionality breaks. That should also point you towards existing tests (which can be removed/refactored/freshly designed as you see fit).